Add full inspector

Sedictious opened the pull request:
Note: I tried to keep this PR focused but it again got slightly out of hand since it can be partitioned into 2 smaller PRs (add the base for the full inspector and the small multi-select refactoring). I can revert if it's preferable :)

This adds the basic full inspector and adds it to the graph designer (I left out the 'delete' button since I felt like this only needs to be an option in the edge/node tool)

Sedictious committed:
Add bottom toolbar

Sedictious committed:
bottomtoolbar: Replace with radio butons

Sedictious committed:
Make toolbar functional

Sedictious committed:
Fixes

Sedictious committed:
Add full inspector

cardinot commented on src/gui/fullinspector.h:
```diff
@@ -0,0 +1,50 @@
+#include <QStringList>
+
+#include "graphdesigner.h"
+#include "graphdesignerpage.h"
```
You've included te `fullinspector.h` file in the `graphdesigner.h` and `graphdesignerpage.h` too. It produces a recursion error.

In any case, since the idea is to reuse the `FullInspector` in the context of an Experiment, this widget should not depend on the `GraphDesignerPage` at all. So, the constructor should be `explicit FullInspector(QWidget* parent);`, and those includes could be removed

cardinot commented on src/gui/basegraphgl.cpp:
```diff
@@ -381,6 +391,34 @@ void BaseGraphGL::mouseReleaseEvent(QMouseEvent *e)
 }
 
+
+void BaseGraphGL::updateFullInspector()
```
as mentioned before... the `BaseGraphGL` should be kept independent of the inspectors.

That is, this class should not be changing anything in the `FullInspector` because it'd introduce a flaw in the architecture. Instead, it could just emit a signal to let the listeners know when something was selected. Then, the listeners should do the job of making the FullInspector aware that its contents should be updated.

Note that `BaseGraphGL` is instantiated by the `GraphWidget` class, which is instantiated by the `GraphDesigner` class. The `GraphDesigner` in turn is instantiated by the `GraphDesignerPage` class, which also instantiates the `FullInspector` , i.e.,

`GraphDesignerPage` -> `GraphDesigner` -> `GraphWidget` -> `BaseGraphGL`

`GraphDesignerPage` -> `FullInspector`

Thus,
- the `FullInspector` could have a public slot like `void slotSelectedNode(const Node& node);` with the logic to update itself with the information provided by the selected node
- the `BaseGraphGL` could emit a `nodeSelectionChanged(node);`
- the `GraphDesignerPage` could make the connection between them

Sedictious committed:
Refactor full inspector

Sedictious committed:
Full inspector is independent from basegraphgl

Sedictious committed:
Merge branch 'socis19' into fullinspector

Sedictious committed:
Minor fixes

cardinot commented on src/gui/graphdesignerpage.h:
```diff
@@ -25,19 +25,21 @@
-
-#include "core/mainapp.h"
-#include "core/graphinputs.h"
+#include "graphdesigner.h"
```
recursive include:
`graphdesigner.h` includes `graphdesignerpage.h`; and `graphdesignerpage.h` includes `graphdesigner.h`

In any case, you already have a forward declaration of `GraphDesigner` at L41. So, you could remove L34 and add the include in the cpp file

cardinot commented on src/gui/graphdesignerpage.h:
```diff
@@ -62,12 +64,15 @@ class GraphDesignerPage : public QMainWindow
     inline QStringList graphAttrHeader() const;
     inline QStringList graphAttrValues() const;
     inline PluginKey selectedGraphKey() const;
+    inline FullInspector* fullInspector();
```
getter functions should be `const`, i.e.,
`inline FullInspector* fullInspector() const;`

cardinot commented on src/gui/graphdesignerpage.h:
```diff
@@ -89,6 +94,9 @@ private slots:
     void openSettingsDlg();
 };
 
+inline FullInspector* GraphDesignerPage::fullInspector(){
```
coding style?

cardinot commented on src/gui/graphview.cpp:
```diff
@@ -30,12 +30,12 @@ namespace evoplex {
+     m_edgeCMap(nullptr),
+     m_edgeScale(25.),
+     m_edgePen(Qt::gray),
+     m_nodePen(Qt::black)
```
please, don't break the coding style (members should be lined up)
``` cpp
MyClass::MyClass()
    : abc(1),
      qwe(2),
      qwert(2)
{
}
```

cardinot commented on src/gui/graphwidget.cpp:
```diff
@@ -115,6 +115,10 @@ void GraphWidget::changeSelectionMode(SelectionMode m) {
     m_view->setCurrentSelectionMode(m);
 }
 
+void GraphWidget::setViewInspector(FullInspector* inspector) {
```
why do we need to tie the `FullInspector` with the `GraphWidget`?

cardinot commented on src/gui/basegraphgl.h:
```diff
@@ -80,6 +83,9 @@ class BaseGraphGL : public QOpenGLWidget, public GraphGLInterface
     inline int currStep() const { return m_currStep; }
+    inline void setInspector(FullInspector* inspector);
+    std::map<int, Node> m_selectedNodes; //TODO: This shouldn't be public
+//    void updateFullInspector();
```
I see the TODO note, but we should never ever push code with public members like that...
In any case, I see that `m_selectedNodes` is not being used anyway, so I think it could be removed

cardinot commented on src/gui/basegraphgl.h:
```diff
@@ -145,6 +153,8 @@ private slots:
     QRect m_inspGeo; // inspector geometry with margin
     SelectionMode m_curMode;
     std::vector<std::shared_ptr<AttrWidget>> m_attrWidgets;
+    FullInspector* m_inspector;
```
`BaseGraphGL` should be independent of the `FullInspector`

cardinot commented on src/gui/graphdesigner.cpp:
```diff
@@ -63,6 +63,11 @@ GraphDesigner::GraphDesigner(MainGUI* mainGUI, QWidget* parent)
     layout->addWidget(m_innerWindow);
+    
+    m_curGraph->setViewInspector(parent->fullInspector());
+    connect(m_curGraph->view(), SIGNAL(nodeSelected(const Node&)), parent->fullInspector(), SLOT(slotSelectedNode(const Node&)));
```
https://github.com/evoplex/evoplex/pull/44#discussion_r310407646
https://github.com/evoplex/evoplex/pull/44#discussion_r310407646
"the GraphDesignerPage could make the connection between them"

cardinot commented on src/gui/graphdesigner.cpp:
```diff
@@ -42,7 +42,7 @@
 namespace evoplex {
 
-GraphDesigner::GraphDesigner(MainGUI* mainGUI, QWidget* parent)
+GraphDesigner::GraphDesigner(MainGUI* mainGUI, GraphDesignerPage* parent)
```
We should not make this class strongly tied to the `GraphDesignerPage`. Please, revert this line to take a `QWidget` instead of the `GraphDesignerPage`

Sedictious committed:
Basegraphgl doesn't depend to fullinspector

Sedictious commented:
@cardinot Those were all bad design choices stemming from the fact that I decided to make the FullInspector accessible from the BaseGraphGL. I refactored the code and I think I now addressed all of your comments

cardinot commented on src/gui/basegraphgl.cpp:
```diff
@@ -1,22 +1,22 @@
+*
+*  You should have received a copy of the GNU General Public License
+*  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
```
please, keep it consistent with the other files, i.e.,
``` c++
\*
 *
 *\
```
(please, do the same for the fullinspector files)

cardinot commented on src/gui/basegraphgl.h:
```diff
@@ -80,6 +83,7 @@ class BaseGraphGL : public QOpenGLWidget, public GraphGLInterface
     void paint(QPaintDevice* device, bool paintBackground) const;
 
     inline int currStep() const { return m_currStep; }
+    inline void setInspector(FullInspector* inspector);
```
unused?

cardinot commented on src/gui/basegraphgl.h:
```diff
@@ -145,6 +151,8 @@ private slots:
     SelectionMode m_curMode;
     std::vector<std::shared_ptr<AttrWidget>> m_attrWidgets;
+    std::map<int, Node> m_selectedNodes;
+    bool m_fullInspectorVisible;
```
`m_selectedNodes` and `m_fullInspectorVisible` are not being used

cardinot commented on src/gui/basegraphgl.h:
```diff
@@ -21,6 +21,7 @@
 #ifndef BASEGRAPHGL_H
 #define BASEGRAPHGL_H
 
+#include <map>
```
this can be removed

cardinot commented on src/gui/graphdesignerpage.cpp:
```diff
@@ -64,8 +65,15 @@ GraphDesignerPage::GraphDesignerPage(MainGUI* mainGUI)
     connect(m_ui->acEdgeTool, &QAction::triggered, [this]() { this->m_graphDesigner->slotChangeSelectionMode(SelectionMode::EdgeEdit); });
+    
+    connect(m_graphDesigner->graphView(), SIGNAL(nodeSelected(const Node&)), fullInspector(), SLOT(slotSelectedNode(const Node&)));
+    connect(m_graphDesigner->graphView(), SIGNAL(clearedSelected()), fullInspector(), SLOT(slotClear()));
```
m_inspector instead of fullInspector()

cardinot commented on src/gui/graphdesignerpage.cpp:
```diff
@@ -64,8 +65,15 @@ GraphDesignerPage::GraphDesignerPage(MainGUI* mainGUI)
     connect(m_ui->acNodeTool, &QAction::triggered, [this]() { this->m_graphDesigner->slotChangeSelectionMode(SelectionMode::NodeEdit); });
     connect(m_ui->acEdgeTool, &QAction::triggered, [this]() { this->m_graphDesigner->slotChangeSelectionMode(SelectionMode::EdgeEdit); });
+    
+    connect(m_graphDesigner->graphView(), SIGNAL(nodeSelected(const Node&)), fullInspector(), SLOT(slotSelectedNode(const Node&)));
```
m_inspector instead of fullInspector()

cardinot commented on src/gui/graphdesignerpage.h:
```diff
@@ -25,19 +25,20 @@
-
-#include "core/mainapp.h"
-#include "core/graphinputs.h"
+#include "fullinspector.h"
```
you should add a forward declaration OR include the header
in thise case, you could move this `#include "fullinspector.h"` to the cpp file

cardinot commented on src/gui/graphdesignerpage.h:
```diff
@@ -25,19 +25,20 @@
 
 #include "maingui.h"
 #include "abstractgraph.h"
 #include "attributerange.h"
```
``` c++
#include "core/include/abstractgraph.h"
#include "core/include/attributerange.h"
```
(you can place them together with the other lines from core/include)

cardinot commented on src/gui/graphview.cpp:
```diff
@@ -3,7 +3,7 @@
  *  Evoplex is a multi-agent system for networks.
  *  Copyright (C) 2018 - Marcos Cardinot <marcos@cardinot.net>
- *
+ * 
```
?

cardinot commented on src/gui/graphwidget.h:
```diff
@@ -33,6 +33,7 @@ namespace evoplex {
 
 class BaseGraphGL;
 class GraphWidget;
+enum class SelectionMode;
```
since the file already includes the `#include "basegraphgl.h"`, the lines 34-36 could be removed

cardinot commented on src/gui/fullinspector.cpp:
```diff
@@ -0,0 +1,54 @@
+FullInspector::FullInspector(QWidget* parent)
+    : QDockWidget(parent),
+    m_ui(new Ui_FullInspector),
+    m_parent(parent)
```
coding style

Sedictious committed:
Formatting and minor fixes

Sedictious committed:
Full Inspector remains open

Sedictious committed:
Add inspector mode toggle

Sedictious committed:
Display attributes

Sedictious committed:
fullinspector: Edit attributes

Sedictious committed:
Fix layout

Sedictious committed:
Fix errors

Sedictious committed:
Intendation fix

Sedictious committed:
Add check for empty graph inputs

Sedictious committed:
Don't parse inputs before initializing a graph
